File: Algorithms/C++/Backtracking/77_Combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> combinations;

        if (n <= 0 || k < 0 || k > n) {
            return combinations;
        }

        vector<int> nums;
        for (int i = 1; i <= n; i++) {
            nums.push_back(i);
        }

        vector<int> cur;

        helper(combinations, cur, 0, nums, k);

        return combinations;
    }
private:
    void helper(vector<vector<int>>& combinations, vector<int>& cur, int pos, vector<int> nums, int k) {
        if (k == 0) {
            combinations.push_back(cur);
            return;
        }

        for (int i = pos; i < nums.size(); i++) {
            cur.push_back(nums[i]);
            helper(combinations, cur, i + 1, nums, k - 1);
            cur.pop_back();
        }
    }
};

File: Algorithms/Swift/Backtracking/77_Combinations.swift
class Solution {
    func combine(n: Int, _ k: Int) -> [[Int]] {
        var combinations = [[Int]]()

        if n <= 0 || n < k {
            return combinations
        }

        var nums = [Int]()
        for i in 1...n {
            nums.append(i)
        }

        var currentCombination = [Int]()
        helper(&combinations, &currentCombination, 0, nums, k)

        return combinations
    }

    private func helper(inout combinations: [[Int]], inout _ currentCombination: [Int], _ position: Int, _ nums: [Int], _ k: Int) {
        if k == 0 {
            combinations.append(currentCombination)
            return
        }

        for var i = position; i < nums.count; i++ {
            currentCombination.append(nums[i])
            helper(&combinations, &currentCombination, i + 1, nums, k - 1)
            currentCombination.removeLast()
        }
    }
}
